lib/writeGltf.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+var fs = require('fs');
+var path = require('path');
+var Cesium = require('cesium');
+var defined = Cesium.defined;
+var dataUri = require('datauri');
+var async = require('async');
+var mkdirp = require('mkdirp');
+
+module.exports = writeGltf;
+
+function writeGltf(gltf, outputPath, isEmbedded, createDirectory, callback) {
+    //Create the output directory if specified
+    if (createDirectory) {
+        outputPath = path.join(path.dirname(outputPath), 'output', path.basename(outputPath));
+        mkdirp.sync(path.dirname(outputPath));
+    }
+    
+    var shaders = gltf.shaders;
+    //Write out the source objects, followed by the entire gltf object
+    if (defined(shaders)) {
+        var shaderIds = Object.keys(shaders);
+        async.each(shaderIds, function(shaderId, asyncCallback) {
+            var shader = shaders[shaderId];
+            if (defined(shader.extras) && defined(shader.extras.source)) {
+                var source = shader.extras.source;
+                delete shader.extras.source;
+
+                var filePath = path.dirname(outputPath);
+                writeSource(filePath, shaderId, shader, source, isEmbedded, asyncCallback);
+            }
+        }, function(err) {
+            if (err) {
+                if (callback) {
+                        process.nextTick(function() {

```

Comment:
> When you first commented in the previous PR, it was for a callback within readFile.

Sorry, I was wrong there.

> (In other words, I would only use process.nextTick to make the data URI case asynchronous, right?)

Exactly because you want the callback to appear to be async.